src/type/matrix/utils/broadcast.js

```diff
@@ -0,0 +1,84 @@
+import { factory } from '../../../utils/factory.js'
+
+const name = 'broadcast'
+
+const dependancies = ['concat']
+
+export const createBroadcast = /* #__PURE__ */ factory(
+  name, dependancies,
+  ({ concat }) => {
+    /**
+   * Broadcasts two matrices, and return both in an array
+   * It checks if it's possible with broadcasting rules
+   *
+   * @param {Matrix}   A      First Matrix
+   * @param {Matrix}   B      Second Matrix
+   *
+   * @return {Matrix[]}      [ broadcastedA, broadcastedB ]
+   */
+    return (A, B) => {
+      let sizeA
+      let sizeB
+      const N = Math.max(A._size.length, B._size.length) // max number of dims
+      if (A._size.length === B._size.length) {
+        if (A._size.every((dim, i) => dim === B._size[i])) {
+          // If matrices have the same size return them
+          return [A, B]
+        }
+        sizeA = [...A._size] // clone
+        sizeB = [...B._size] // clone
+      } else {
+        sizeA = A._size.length < N
+          ? _padLeft(A._size, N, 0) // pad to the left to align dimensions to the right
+          : [...A._size] // clone
+
+        sizeB = B._size.length < N
+          ? _padLeft(B._size, N, 0) // pad to the left to align dimensions to the right
+          : [...B._size] // clone
+      }
+
+      // calculate the max dimensions
+      const sizeMax = []
+
+      for (let dim = 0; dim < N; dim++) {
+        sizeMax[dim] = Math.max(sizeA[dim], sizeB[dim])
+      }
+
+      // check if the broadcasting rules applyes for both matrices
+      for (let dim = 0; dim < N; dim++) {
+        _checkRules(sizeA, sizeMax, dim)
+        _checkRules(sizeB, sizeMax, dim)
+      }
+
+      // reshape A or B if needed to make them ready for concat
+      if (A._size.length < N) {
+        A.reshape(_padLeft(A._size, N, 1))
```

Comment:
`A.reshape` and `B.reshape` mutate the original matrix. I think that we should leave the original matrices A and B untouched, and create a clone first. We should also write a test validating that `broadcast` is immutable. What do you think?

Reply:
Since the matrix already gets bigger, I was concerned about also making a clone and keeping both the original matrices and the broadcasted clones, possibly using more memory.

In the end I only included `broadcast` as a function that is called by `matrixAlgorithmSuite` since I didn't find a clean way to extract both values inside the mathjs parser like `[Abroadcasted, Bbroadcasted] = broadcast(A, B)`

In the case it's called from `matrixAlgorithmSuite` it doesn't mutate the input matrices outside the scope `broadcast`

```js
A = [1, 2]
B = [[3], [4]]
math.add(A, B)           // [[4, 5], [5, 6]]
A                        // [1, 2]
B                        // [[3], [4]]
```

Should something like this be included in tests?

It's not much effort to make the clones internally from `broadcast`, It certainly shouldn't mutate the input matrices, I think it might use more memory and doesn't seem to be strictly necessary.

Nonetheless it's your call, in the end the way would be not mutating or cloning the matrices but modifying indexing as if the matrices where stretched in all dimensions (but first checking the rules of broadcasting)

Reply:
Hi Jos, while trying to make broadcast.test I did encounter the mutation you were telling me about. Thus I made clones as suggested and it fixed the problem. I didn't encounter that mutation on the scopes I was manually testing, my bad.

Reply:
My general rule is:
- Functions that return something should not touch their input arguments, i.e. `C = fn(A, B)` should not change A or B.
- If you want to make a change in an input argument, then ideally the API communicates that by returning nothing. So for example `fn(A)` returning nothing indicates that `fn` does make some change in A.

It is a pity that we have to create a clone for broadcast just for internal use, but I think it _very_ important not to change the input arguments. So we have to. Like discussed before, with a different approach more similar to `reduce` we may be able to create a more efficient implementation, but for now, we go for a _simple_ implementation.

> Hi Jos, while making `broadcast.test` I did encounter the mutation you were telling me about. Thus I made clones as suggested and it fixed the problem. I didn't encounter that mutation by testing it from the functions that call broadcast.

That is interesting 🤔 . Only reason that I can think of is that the matrices are already cloned somewhere else? If that is the case, they would be cloned twice. Can you figure out if that is indeed the case? Ideally we should only clone once of course :).

Reply:
In `matrixAlgorithmSuite` I did implement broadcast by changing from `(x, y) => matAlgo13xDD(x, y, elop)` to ` (x, y) => matAlgo13xDD(...broadcast(x, y), elop)`. My intention was to split the output of broadcast that is `[x_broad, y_broad]` to make it as if it was `(x, y) => matAlgo13xDD(x_broad, y_broad, elop)`.

I recently learned that `...` is also a way to clone. So I think there is a side effect as if it was `(x, y) => matAlgo13xDD(x_broad_clone, y_broad_clone, elop)`. My best guess is that the mutation stops outside of `broadcast` because there are no calls to broadcast without `...broadcast`

I was following up examples like `add` and `to` to see obvious signs of cloning inputs but I didn't find anything.

Reply:
Ok clear. So, let's go for the safe option then, just clone the matrices inside broadcast so we are sure we're not mutating anything. The spread operator indeed only does a shallow clone so that would not be enough.